Change GetNumChildren()/CalculateNumChildren() methods return llvm::Expected

adrian-prantl opened the pull request:
Change GetNumChildren()/CalculateNumChildren() methods return llvm::Expected

This is an NFC change that does not yet add any error handling or change any code to return any errors.

This is the second big change in the patch series started with https://github.com/llvm/llvm-project/pull/83501

A follow-up PR will wire up error handling.

walter-erquinigo commented on lldb/source/DataFormatters/FormatManager.cpp:
```diff
@@ -451,8 +451,13 @@ bool FormatManager::ShouldPrintAsOneLiner(ValueObject &valobj) {
 
+  auto num_children = valobj.GetNumChildren();
+  if (!num_children) {
+    llvm::consumeError(num_children.takeError());
```
could you log this some channel instead of consuming the error? That would still be NFC and would fit well within this PR

adrian-prantl replied:
Done!

walter-erquinigo commented on lldb/source/Target/StackFrame.cpp:
```diff
@@ -1397,7 +1398,9 @@ ValueObjectSP GetValueForOffset(StackFrame &frame, ValueObjectSP &parent,
   }
 
-  for (int ci = 0, ce = parent->GetNumChildren(); ci != ce; ++ci) {
+  for (int ci = 0, ce = llvm::expectedToStdOptional(parent->GetNumChildren())
```
I'm afraid that calls to `expectedToStdOptional` will just silence important warnings.

adrian-prantl commented:
@walter-erquinigo I was wondering if it would make sense to create a DropAndLogError() wrapper for all these places. The main problem would be that we end up calling these functions dozens of times over and over so we'd be spamming the logs quite a bit. But maybe doing a verbose log would be appropriate. I'll modify the patch accordingly.

walter-erquinigo commented:
@adrian-prantl , I like that idea. Dump the error messages only to the formatters channel in verbose mode.

JDevlieghere reviewed (changes requested):
I'm worried about making it too easy to ignore errors. The whole point of the llvm::Error class is that you have to check it. There might be a legitimate reason you don't about the error the call site, but it should very explicit and easily auditable. If none of the callers care about the error, then it shouldn't be an Error/Expected in the first place. We have plenty of examples of that in LLVM. 

I totally understand the need for this as a transitional tool, but I want to make sure nobody writes new code using this. I can see two ways to make discourage that:

 - Mark the helper as deprecated. This might generate a lot off noise though. 
 - Make it really obvious in the name that this is a temporary or that we're purposely dropping the error. Either something like `FIX_ERROR_HANDLING` or `LLDB_DROP_ERROR` or something. Logging to the verbose log is better than nothing, but it doesn't constitute error handling. 

PS: The current implementation wraps the `LLDB_LOG_*` which uses #line and #file information. Now all those errors will get attributed to this helper function.

jimingham commented:
I would say this differently.  Many clients of the GetNumChildren API are planning to respond the same way to an error and a number of children == 0.  Having to embed this "i'm not actually checking the error, I'm just converting it to num-children = 0" snippet looks odd, like "Why are you bothering with errors when you just discard them immediately for "value is 0".

A better title for this use of GetNumChildren would be something like GetNumChildrenErrorsAreZero.  That has the virtue of saying what's going on explicitly.

Jim

> On Mar 7, 2024, at 10:45 AM, Jonas Devlieghere ***@***.***> wrote:
> 
> 
> @JDevlieghere requested changes on this pull request.
> 
> I'm worried about making it too easy to ignore errors. The whole point of the llvm::Error class is that you have to check it. There might be a legitimate reason you don't about the error the call site, but it should very explicit and easily auditable. If none of the callers care about the error, then it shouldn't be an Error/Expected in the first place. We have plenty of examples of that in LLVM.
> 
> I totally understand the need for this as a transitional tool, but I want to make sure nobody writes new code using this. I can see two ways to make discourage that:
> 
> Mark the helper as deprecated. This might generate a lot off noise though.
> Make it really obvious in the name that this is a temporary or that we're purposely dropping the error. Either something like FIX_ERROR_HANDLING or LLDB_DROP_ERROR or something. Logging to the verbose log is better than nothing, but it doesn't constitute error handling.
> PS: The current implementation wraps the LLDB_LOG_* which uses #line and #file information. Now all those errors will get attributed to this helper function.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/84219#pullrequestreview-1923251331>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW6WTDMOLUAJSNVWUILYXCYVNAVCNFSM6AAAAABEJS3ASKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMRTGI2TCMZTGE>.
> You are receiving this because you are on a team that was mentioned.
>

adrian-prantl commented:
> I would say this differently.  Many clients of the GetNumChildren API are planning to respond the same way to an error and a number of children == 0.

I haven't done a complete audit, but my expectation for this patch is that the vast majority of users of this API will want to bubble up the error. So I'm seeing this as a transition state, with the goal being to convert all functions that currently use the ValueOr function to also return an Expected.

adrian-prantl committed:
Change GetNumChildren()/CalculateNumChildren() to return llvm::Expected

This is an NFC change that does not yet add any error handling or
change any code to return any errors. Some select APIs may
legitimately not care about errors, so this patch also adds a
GetNumChildrenIgnoringErrors() variant to the API.

As we convert more APIs to returning llvm::Expected, we expect the
number of calls to GetNumChildrenIgnoringErrors() to dwindle.

jimingham commented:
> > I would say this differently.  Many clients of the GetNumChildren API are planning to respond the same way to an error and a number of children == 0.
> 
> I haven't done a complete audit, but my expectation for this patch is that the vast majority of users of this API will want to bubble up the error. So I'm seeing this as a transition state, with the goal being to convert all functions that currently use the ValueOr function to also return an Expected.

Cool!  If that ends up not being the case then it would be good to have something that says the distinction doesn't matter, but if this construct vanishes, all the better!

adrian-prantl commented:
@walter-erquinigo @jimingham I updated the patch to implement Jim's suggestion of adding an explicit `GetNumChilrdrenIgnoringErrors()` API that also ticks off the "big red warning flag" @JDevlieghere was pushing for.

jimingham reviewed (approved):
LGTM.  The "IgnoringErrors" makes this construct easier to spot, and also should give anyone a bad feeling if they are tempted to use it...

fmayer commented:
I think this broke the build completely?

```
/usr/local/google/home/fmayer/large/llvm-project/lldb/source/Plugins/ABI/PowerPC/ABISysV_ppc64.cpp:771:14: error: no viable conversion from 'llvm::Expected<uint32_t>' (aka 'Expected<unsigned int>') to 'uint32_t' (aka 'unsigned int')
    uint32_t n = m_type.GetNumChildren(omit_empty_base_classes, nullptr);
             ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

adrian-prantl commented:
Apologies, I made a last-minute rebase without rebuilding. Thanks for reverting!

adrian-prantl commented:
Or rather, I didn't build with all LLVM targets.
